File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.py
class Solution:
    def genComb(self,digits,vals,ind,temp,ans):
        if ind==len(digits):
            ans.append(temp)
            return
        for i in vals[int(digits[ind])]:
            self.genComb(digits,vals,ind+1,temp+i,ans)

    def letterCombinations(self, digits: str) -> List[str]:
        vals=["","","abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"]
        ans=[]
        if len(digits)==0:
            return ans
        self.genComb(digits,vals,0,"",ans)
        return ans

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
class Solution {
    public void gen(String digits,int index,String[] map,List<String> ans,String s){
        if(index==digits.length()){
            ans.add(s);
            return;
        }
        for(char c: map[digits.charAt(index)-'0'].toCharArray()){
            gen(digits,index+1,map,ans,s+c);
        }
    }
    public List<String> letterCombinations(String digits) {
        String[] map={"","","abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"};
        List<String> ans=new ArrayList<>();
        gen(digits,0,map,ans,"");
        return ans;
    }
}